drivers: hci: stm32_ipm: Add possibility to work on LSI

pavlohamov opened the pull request:
Allow to work on LSI of LSE failed to start (or absent)

Also adding compile time switch for LSE, this nucleo_wb55 board will be extended with 
`
select CLOCK_STM32_LSE
`

Commit:
drivers: hci: stm32_ipm: Add possibility to work on LSI

Use CLOCK_STM32_LSE as switch for the RF low speed clock source

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -327,17 +327,37 @@ static void start_ble_rf(void)
+		LL_RCC_LSE_Enable();
+		while (!LL_RCC_LSE_IsReady()) {
+			if (alarm < k_uptime_get_32())
+				goto init_LSI;
```
I'm fine with adding LSI support (while I have questions around clock is precise enough for RF, I'll checking internally).
Though, I don't think enabling as LSE back up make much sense.
If LSE is available on HW and config LSE is enabled, then it should work. If not working, there is a HW issue to be checked. Can you explain  this choice?

pavlohamov replied:
I'm fine to with busy loop on LSE failure if LSE is configured. 
Keeping device functional even if there is issues with the HW, seems to me a valuable choice. As Far as i know (from my ST contact person) only HSE is required for normal RF operations

erwango replied:
> Keeping device functional even if there is issues with the HW, seems to me a valuable choice.

For a product, it could be useful indeed. Though before this you want to know that you have an issue with your clock. It could be an optonnal feature, but then there is no reason to do it here and not on other places where clock is needed (everywhere).
Please use busy loop on failure.

pavlohamov replied:
Make sense. I'd prefer to have some LOG_ERR for LSE failure and continue to work on LSI. But there is a big change that this error will be just omitted. Done

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -327,17 +327,37 @@ static void start_ble_rf(void)
+		break;
+#endif
+init_LSI:
+		__HAL_RCC_LSI1_ENABLE();
```
I need to cross check that RTC is actually required.
Anyway a comment indicating the match LSI1>RTC, LSI2>RF would be nice

pavlohamov replied:
Wasn't sure about the actual LSI configuration. Datasheet claims that only LSI2 is required for RF, but it's better to be sure. I will try without LSI1 and see how it goes.

pavlohamov replied:
without LSI1 works fine. I've removed it. Comment added

erwango replied:
Thanks, you can remove the LSI1 comment then ;-).
Also, I think we can actually remove the RTC stuff.
Could be a dedicated commit since this is rather a clean up.

pavlohamov replied:
Make sense. Updated usage of LSI, replaced LSI1 -> LSI2

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -327,17 +327,37 @@ static void start_ble_rf(void)
+		LL_RCC_LSI1_Disable();
+		break;
+#endif
+init_LSI:
```
#endif should be located after `init_LSI:`

pavlohamov replied:
indeed. thanks

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -327,17 +327,30 @@ static void start_ble_rf(void)
-	LL_RCC_LSI1_Disable();
+	LL_RCC_LSI2_Disable();
+#else
+	__HAL_RCC_LSI2_ENABLE();
```
For coherency, can you use LL API?

pavlohamov replied:
Sure! Done

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -327,17 +327,30 @@ static void start_ble_rf(void)
+	LL_RCC_SetRTCClockSource(LL_RCC_RTC_CLKSOURCE_LSI);
+
+	/* Switch OFF LSE */
+	__HAL_RCC_LSE_CONFIG(RCC_LSE_OFF);
```
ditto

erwango reviewed (changes requested):
Few tweaks.
Alos, please update the commit message following change of functionality

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -337,7 +338,21 @@ static void start_ble_rf(void)
 	LL_RCC_SetRTCClockSource(LL_RCC_RTC_CLKSOURCE_LSE);
 
 	/* Switch OFF LSI */
-	LL_RCC_LSI1_Disable();
```
Was this a bugfix?

pavlohamov replied:
Got your point. Yes, actually LSI_1 is used by RTC, but LSI2 is used by RF. Thus there is no sense in touching LSI1 in BLE. Moved to separate commit

Commit:
drivers: hci: stm32_ipm: disable proper LSI

LSI1 is used for RTC, LTC2 for RF.
Disable LSI2 instead of LSI1 on BLE with LSE start

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

erwango commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -338,6 +339,20 @@ static void start_ble_rf(void)
+
+	/* Switch OFF LSE */
+	LL_RCC_LSE_Disable();
+	LL_RCC_LSE_DisableBypass();
```
I don't this it is needed to switch off LSE in this case anymore.
Hasen't been enabled here prevously, so if enabled, it might be by other component that needs it

pavlohamov replied:
Actually, you're right! I was thinking about managing the LSE in term if power consumption. But it's not BLE driver responsibility. LSE operations are deleted. Thank you pushing this idea

erwango reviewed (changes requested):
Please also update commit message for fisrt commit (no fall back scenario anymore)

erwango reviewed (changes requested):
Still needs to update the message for first commit, otherwise LGTM

pavlohamov commented:
> 
> 
> Still needs to update the message for first commit, otherwise LGTM

D'oh, Done
